### PR TITLE
Samples for enhanced enums feature

### PR DESCRIPTION
**Pull request by miquelbeltran**
Hi! :wave: 

I'm doing some work for @RedBrogdon, and creating samples for some new upcoming Dart features.

In this PR, I have created a series of examples for the upcoming enhanced enums feature https://github.com/dart-lang/sdk/issues/47849

I followed the same pattern as in the `extension_methods` examples, creating different files showing different functionality.

I also have created a complete example inside the `complete_example.dart` file.

All examples also contain tests.

### CODE REVIEW

File: enhanced_enums/pubspec.lock [hunk 1]
```diff
@@ -0,0 +1,334 @@
+# Generated by pub
```
Comment (kevmoo) [inline enhanced_enums/pubspec.lock:1]: We shouldn't need this, right?

Comment (RedBrogdon) [reply enhanced_enums/pubspec.lock:1]: The lock file in general, you mean?

Comment (RedBrogdon) [reply enhanced_enums/pubspec.lock:1]: The docs agree with you:

https://dart.dev/guides/libraries/private-files

As a person who arrived at Dart through Flutter, I'm used to checking in the lock file for pretty much everything, whereas you and Parker (as true Dart coders) are probably used to leaving it out more often than including it.

Looks like you're correct, though, so +1 for yanking it. 😄

Comment (domesticmouse) [reply enhanced_enums/pubspec.lock:1]: That guideline is for libraries, and this, unless my reading is incorrect, isn't a library. 

My understanding is we check in `pubspec.lock` files for executable samples as a way of making it really easy for people to get a working version of the code. `dart pub get` will pull a tested set of dependencies that we know work together.

Please revert.

Comment (johnpryan) [reply enhanced_enums/pubspec.lock:1]: Right now, the convention in this repo is *not* to commit the lockfiles.

Comment (miquelbeltran) [reply enhanced_enums/pubspec.lock]: Thanks! I will follow the repo convention then and remove the file.

**Review by johnpryan — CHANGES_REQUESTED** [review]
Looks good overall, I would consider demonstrating that `==` and `hashCode` aren't required and that switch/case can be used with enhanced enums too.

**Review by johnpryan — APPROVED** [review]

**Comment by domesticmouse** [comment]
Does this repository not have a CI? My reading of this PR with it's minimum Dart version should fail when Dart stable is run.

**Comment by johnpryan** [comment]
CI won't run until we do the mono_repo dance, I'm okay with merging into main and creating a beta branch if we don't have one already.

**Review by domesticmouse — APPROVED** [review]

**Pull request merged**